doc/man/man3/uart.3
.TH "UART" 3 "Wed Mar 21 2012" "rfLPC" \" -*- nroff -*-
.ad l
.nh
.SH NAME
UART \- 
.PP
This is the driver for the UART0/2/3\&.  

.SS "Enumerations"

.in +1c
.ti -1c
.RI "enum \fBrflpc_uart_t\fP { \fBRFLPC_UART0\fP =  0, \fBRFLPC_UART2\fP =  2, \fBRFLPC_UART3\fP =  3 }"
.br
.RI "\fIConstants that represents the UART port you want to use in uart driver functions\&. \fP"
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "int \fBrflpc_uart_init\fP (\fBrflpc_uart_t\fP uart_num)"
.br
.RI "\fIInits the UART e, using 11520 baud, 8 bits data, no parity and 1 stop bit\&. \fP"
.ti -1c
.RI "void \fBrflpc_uart_putchar\fP (\fBrflpc_uart_t\fP uart_num, char c)"
.br
.RI "\fISends a byte to the uart\&. \fP"
.ti -1c
.RI "int \fBrflpc_uart_byte_available\fP (\fBrflpc_uart_t\fP uart_num)"
.br
.RI "\fITest the reception FIFO\&. \fP"
.ti -1c
.RI "char \fBrflpc_uart_getchar\fP (\fBrflpc_uart_t\fP uart_num)"
.br
.RI "\fIGets a character from the uart\&. \fP"
.ti -1c
.RI "void \fBrflpc_uart_set_rx_callback\fP (\fBrflpc_uart_t\fP uart_num, \fBrflpc_irq_handler_t\fP callback)"
.br
.RI "\fISets the uart rx callback\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
This is the driver for the UART0/2/3\&. 

At the moment, this driver is only usable when the CPU is clocked at 96Mhz and the UART mode is set to
.IP "\(bu" 2
115200 bauds
.IP "\(bu" 2
8 bits data
.IP "\(bu" 2
no parity check
.IP "\(bu" 2
1 stop bit 
.PP
\fBTodo\fP
.RS 4
Allow fine configuration for the UART\&. Just have to calculate divider and multiplier depending on cpu clock and desired speed\&.
.RE
.PP

.PP

.SH "Enumeration Type Documentation"
.PP 
.SS "enum \fBrflpc_uart_t\fP"

.PP
Constants that represents the UART port you want to use in uart driver functions\&. 
.PP
\fBEnumerator: \fP
.in +1c
.TP
\fB\fIRFLPC_UART0 \fP\fP
Use UART0\&. 
.TP
\fB\fIRFLPC_UART2 \fP\fP
Use UART2\&. 
.TP
\fB\fIRFLPC_UART3 \fP\fP
Use UART3\&. 
.PP
Definition at line 45 of file uart\&.h\&.
.SH "Function Documentation"
.PP 
.SS "int \fBrflpc_uart_byte_available\fP (\fBrflpc_uart_t\fPuart_num)"

.PP
Test the reception FIFO\&. \fBParameters:\fP
.RS 4
\fIuart_num\fP the uart to use 
.RE
.PP
\fBReturns:\fP
.RS 4
true if a byte is available 
.RE
.PP

.SS "char \fBrflpc_uart_getchar\fP (\fBrflpc_uart_t\fPuart_num)"

.PP
Gets a character from the uart\&. \fBParameters:\fP
.RS 4
\fIuart_num\fP the uart to use 
.RE
.PP
\fBReturns:\fP
.RS 4
the character read\&. 
.RE
.PP
\fBWarning:\fP
.RS 4
This function loops while no data is available! 
.RE
.PP

.SS "int \fBrflpc_uart_init\fP (\fBrflpc_uart_t\fPuart_num)"

.PP
Inits the UART e, using 11520 baud, 8 bits data, no parity and 1 stop bit\&. \fBReturns:\fP
.RS 4
0 if init is successful, -1 otherwise 
.RE
.PP
\fBNote:\fP
.RS 4
at the moment, -1 is returned if the CPU is not clocked at 96Mhz 
.RE
.PP

.SS "void \fBrflpc_uart_putchar\fP (\fBrflpc_uart_t\fPuart_num, charc)"

.PP
Sends a byte to the uart\&. \fBParameters:\fP
.RS 4
\fIuart_num\fP the uart to use 
.br
\fIc\fP character to send 
.RE
.PP

.SS "void \fBrflpc_uart_set_rx_callback\fP (\fBrflpc_uart_t\fPuart_num, \fBrflpc_irq_handler_t\fPcallback)"

.PP
Sets the uart rx callback\&. This enables the uart0 interrupt and set the handler accordingly 
.PP
\fBParameters:\fP
.RS 4
\fIuart_num\fP the uart to use 
.br
\fIcallback\fP a pointer to a \fBrflpc_irq_handler_t\fP function 
.RE
.PP

.SH "Author"
.PP 
Generated automatically by Doxygen for rfLPC from the source code\&.


rflpc17xx/drivers/uart.c
/* This file is part of rflpc. Copyright 2010-2011 Michael Hauspie                        
 *									 
 * rflpc is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by	 
 * the Free Software Foundation, either version 3 of the License, or	 
 * (at your option) any later version.					 
 * 									 
 * rflpc is distributed in the hope that it will be useful,		 
 * but WITHOUT ANY WARRANTY; without even the implied warranty of	 
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the	 
 * GNU General Public License for more details.				 
 * 									 
 * You should have received a copy of the GNU General Public License	 
 * along with rflpc.  If not, see <http://www.gnu.org/licenses/>.	 
 */

/*
  Author: Michael Hauspie <michael.hauspie@univ-lille1.fr>
  Contributor: Francois Serman <francois.serman@inria.fr>
*/

#ifdef RFLPC_CONFIG_ENABLE_UART

#include <stdint.h>
#include <stdlib.h>
#include "../clock.h"
#include "../interrupt.h"
#include "../pinconf.h"
#include "../debug.h"

#include "uart.h"

struct uart_pin_conf
{
  /* (p. 108 and p. 299) */
  struct {
    unsigned char port:2;
    unsigned char tx_pin:5;
    unsigned char rx_pin:5;
    unsigned char pin_function:2;
  }gpio;  

  LPC_UART_TypeDef *base_address;
};

typedef struct uart_pin_conf uart_pin_conf_t;

static uart_pin_conf_t _rflpc_uart_config[4] = 
  {
    { 
      { RFLPC_UART0_PORT, RFLPC_UART0_TXD_PIN, RFLPC_UART0_RXD_PIN, RFLPC_UART0_PIN_FUNCTION },
      (LPC_UART_TypeDef*)LPC_UART0,
    },
    { 
      { 0, 0, 0, 0 },      
      NULL,
    },
    { 
      { RFLPC_UART2_PORT, RFLPC_UART2_TXD_PIN, RFLPC_UART2_RXD_PIN, RFLPC_UART2_PIN_FUNCTION },
      LPC_UART2,
    },
    { 
      { RFLPC_UART3_PORT, RFLPC_UART3_TXD_PIN, RFLPC_UART3_RXD_PIN, RFLPC_UART3_PIN_FUNCTION },
      LPC_UART3,
    }
      
  };


#define BASE_ADDR(n) (_rflpc_uart_config[(n)].base_address)

void rflpc_uart_putchar(rflpc_uart_t uart_num, char c)
{
  while (!(BASE_ADDR(uart_num)->LSR & (0x1UL << 5))); /* why << 5 here and not in available? */
  BASE_ADDR(uart_num)->THR = c & 0xFF;
}

int rflpc_uart_byte_available(rflpc_uart_t uart_num)
{
  return BASE_ADDR(uart_num)->LSR & 0x1UL;
}

char rflpc_uart_getchar(rflpc_uart_t uart_num)
{
  while (!rflpc_uart_byte_available(uart_num));
  return BASE_ADDR(uart_num)->RBR & 0xFF;
}


/* For now, we suppose that the CCLK is 96 Mhz. Thus,
   we set PCLK to 12 Mhz by setting it to CCLK/8.
   Then, (p. 315) we have to set 
   DIVADDVAL = 5
   MULVAL = 8
   DLM = 0
   DLL = 4
   UART rate will then be set to 115384
*/

int _rflpc_uart_init(uart_pin_conf_t *uart)
{
   
  /* Set line control values (p. 306) */
  /* 8 data bits, no parity, 1 stopbit */
  /* 3 is for 8 data bits, 7 is to enable DLL and DLM registers (for selecting baud rate) */
  uart->base_address->LCR = (3 | (1 << 7));

  /* Set DLL=4, DLM=0 (p. 315) */
  uart->base_address->DLL = 4;
  uart->base_address->DLM = 0;

  /* Set DIVADDVAL = 5 and MULVAL = 8 (p. 315) */
  uart->base_address->FDR = 5 | (8 << 4);

  /* Enable Fifo and clear TX & RX (p. 305) */
  uart->base_address->FCR = 7;

  /* Set pin mode for the UART to UART0 (TXD0, RXD0)  */
  /* CHANGEME */
  rflpc_pin_set(uart->gpio.port, uart->gpio.rx_pin, uart->gpio.pin_function, 0, 0); /** @todo function value bad for UART3 */
  rflpc_pin_set(uart->gpio.port, uart->gpio.tx_pin, uart->gpio.pin_function, 0, 0);

  /* Reset the DLAB bit in U0LCR to enable access to transmit and receive registers (p. 301) */
  /* CHECKME */
  uart->base_address->LCR &= ~(0x1UL << 7);

  return 0;
}

int rflpc_uart_init(rflpc_uart_t uart_num)
{

  /* Set the peripheral clock to 12 Mhz */
  /* First, check if system clock is 96 Mhz */
  if (rflpc_clock_get_system_clock() != 96000000)
    return -1;


  switch (uart_num)
  {
    /* Set UART CLOCK to 12 Mhz */
    /* 12 Mhz because of the frequency at 96 Mhz */
    case RFLPC_UART0:
      /* Enable UART (user manual, p. 63) */
      LPC_SC->PCONP |= (1UL << 3);
      /* Set UART clock Bits 6 and 7 are for UART3, 0x3 value is for CCLK/8 */
      LPC_SC->PCLKSEL0 |= (RFLPC_CCLK_8 << 6);
      break;
    case RFLPC_UART2:
      /* Enable UART (user manual, p. 63) */
      LPC_SC->PCONP |= (1UL << 24);
      /* Bits 16 and 17 are for UART3, 0x3 value is for CCLK/8 */
      LPC_SC->PCLKSEL1 |= (RFLPC_CCLK_8 << 16);
      break;
    case RFLPC_UART3:
      /* Enable UART (user manual, p. 63) */
      LPC_SC->PCONP |= (1UL << 25);
      /* Bits 18 and 19 are for UART3, 0x3 value is for CCLK/8 */
      LPC_SC->PCLKSEL1 |= (RFLPC_CCLK_8 << 18);
      break;
    default:
      return -1;
  }

  return _rflpc_uart_init(&_rflpc_uart_config[uart_num]);
}



void rflpc_uart_set_rx_callback(rflpc_uart_t uart_num, rflpc_irq_handler_t  callback)
{
    RFLPC_ASSERT(uart_num != 1);
    /* set the uart0 interrupt handler */
    rflpc_irq_set_handler(UART0_IRQn + uart_num, callback);
    /* enable the interrupt vector */
    rflpc_irq_enable(UART0_IRQn + uart_num);
    /* enable the uart0 irq generation (user manual p. 302) */
    BASE_ADDR(uart_num)->IER |= 1;
}

#endif /* ENABLE_UART */
